handler/oauth2/flow_authorize_jwt_token.go

```diff
@@ -0,0 +1,330 @@
+	"github.com/ory/x/errorsx"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+const grantTypeJwtBearer = "urn:ietf:params:oauth:grant-type:jwt-bearer"
+
+type AuthorizeJwtGrantHandler struct {
+	AuthorizeJwtGrantStorage AuthorizeJwtGrantStorage
+	ScopeStrategy            fosite.ScopeStrategy
+	AudienceMatchingStrategy fosite.AudienceMatchingStrategy
+
+	// TokenURL is the the URL of the Authorization Server's Token Endpoint.
+	TokenURL string
+	// SkipClientAuth indicates, if client authentication can be skipped.
+	SkipClientAuth bool
+	// JWTIDOptional indicates, if jti (JWT ID) claim required or not.
+	JWTIDOptional bool
+	// JWTIssuedDateOptional indicates, if "iat" (issued at) claim required or not.
+	JWTIssuedDateOptional bool
+	// JWTMaxDuration sets the maximum time after token issued date (if present), during which the token is
+	// considered valid. If "iat" claim is not present, then current time will be used as issued date.
+	JWTMaxDuration time.Duration
+
+	*HandleHelper
+}
+
+// HandleTokenEndpointRequest implements https://tools.ietf.org/html/rfc6749#section-4.1.3 (everything) and
+// https://tools.ietf.org/html/rfc7523#section-2.1 (everything)
+func (c *AuthorizeJwtGrantHandler) HandleTokenEndpointRequest(ctx context.Context, request fosite.AccessRequester) error {
+	if !c.CanHandleTokenEndpointRequest(request) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	// if client is authenticated, check grant types
+	if !c.CanSkipClientAuth(request) && !request.GetClient().GetGrantTypes().Has(grantTypeJwtBearer) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHintf("The OAuth 2.0 Client is not allowed to use authorization grant \"%s\".", grantTypeJwtBearer))
+	}
+
+	assertion := request.GetRequestForm().Get("assertion")
+	if assertion == "" {
+		return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("The assertion request parameter must be set when using grant_type of '%s'.", grantTypeJwtBearer))
+	}
+
+	token, err := jwt.ParseSigned(assertion)
+	if err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("Unable to parse jwt token passed in \"assertion\" request parameter.").
+			WithWrap(err).WithDebug(err.Error()),
+		)
+	}
+
+	// check fo required claims in token, so we can later find public key based on them
+	if err := c.validateTokenPreRequisites(token); err != nil {
+		return err
+	}
+
+	key, err := c.findPublicKeyForToken(ctx, token)
+	if err != nil {
+		return err
+	}
+
+	if err := c.validateToken(ctx, token, key); err != nil {
+		return err
+	}
+
+	claims := jwt.Claims{}
+	if err := token.Claims(key, &claims); err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("Unable to verify the integrity of the 'assertion' value.").
+			WithWrap(err).WithDebug(err.Error()),
+		)
+	}
+
+	if claims.ID != "" {
+		if err := c.AuthorizeJwtGrantStorage.MarkJWTUsedForTime(ctx, claims.ID, claims.Expiry.Time()); err != nil {
```

Comment:
It is better to move this check to the end, otherwise the token is marked as used despite the results of the checks below.